### DISCUSSION TITLE
Project structure

### DISCUSSION THREAD

**Discussion by 5wattbulb** [mesonbuild/meson#9770] [Q&A] [answered]
I think I'm right in saying that on github you'll usually see the src and build directories in the project root directory. Having read through the meson documentation I'm unsure as to the best directory structure to use with meson. I came across this thread
https://stackoverflow.com/questions/55107396/meson-targets-that-depend-on-subdir-siblings
and I was thinking I'd use the same method as suggested in the answer by rellenberg so that all the binaries end up in the build directory rather than subdirs of the build directory. The simple meson examples all seem to have the source code in the project root directory but I'm thinking that's just for convenience rather than a suggestion. Anyone got any views?

Edit: Eli's answer in the "Building progams without needing to specify libraries and source files?" makes things easier regardless of project structure.

**Comment by andy5995** [comment, marked as answer]
@5wattbulb I think it depends on the type of project. If you only have a few source files, having them in root directory is probably ok. For [rmw](https://github.com/theimpossibleastronaut/rmw), I have them all in a separate src/ directory, but the binary is output to the build root directory.

>The simple meson examples all seem to have the source code in the project root directory but I'm thinking that's just for convenience rather than a suggestion.

Yes, probably easier for examples.

To get ideas about project structure, I'd suggest having a look at https://mesonbuild.com/Users.html and see how other projects have things organized.

**Reply by 5wattbulb** [reply]
Thanks Andy. I posted that before I realised you could have an additional meson.build file in the src/ directory containing only
SrcFiles = files('src1.c', 'src2.c', ....
and then use subdir('src') along with the SrcFiles variable in the root meson.build.
Will take a look at that link as I'm badly in need of some examples. I've got to say meson, while still a lot to take in, is a breath of fresh air compared to autotools.

**Reply by andy5995** [reply]
> I've got to say meson, while still a lot to take in, is a breath of fresh air compared to autotools.

I agree. I used autotools for a while, and tried cmake for a (very) short time. I've been using meson now for several months and am quite satisfied with it so far.

**Reply by 5wattbulb** [reply]
I only started learning linux about 9 months ago. Bash was bad enough but when I got to make I went into wall staring mode. Took me ages to work through the manual, repeated twice more and still felt a distance away from using it proficiently. When I looked at the autoconf manual I spent a couple of weeks becoming proficient in M4, returned to the manual and by the time I was a third of the way through it I was losing the will to live. I then came across an article singing the praises of meson and now feel as though I'm on the other side of the brick wall.
